sw/device/lib/dif/dif_plic.c

```diff
@@ -0,0 +1,317 @@
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+// Interrupt Pending register offsets from the PLIC base
+#define RV_PLIC_IP0 0x0u
+#define RV_PLIC_IP1 0x4u
+
+// Level/Edge register offsets from the PLIC base
+#define RV_PLIC_LE0 0x8u
+#define RV_PLIC_LE1 0xCu
+
+// Priority register offsets from the PLIC base
+#define RV_PLIC_PRIO0 0x10u
+
+// Interrupt Enable register offsets from the PLIC base
+#define RV_PLIC_IE00 0x200u
+#define RV_PLIC_IE01 0x204u
+
+// Threshold register offsets from the PLIC base
+#define RV_PLIC_THRESHOLD0 0x208u
+
+// Interrupt Claim register offsets from the PLIC base
+#define RV_PLIC_CC0 0x20cu
+
+// The highest interrupt priority
+#define RV_PLIC_MAX_PRIORITY 0x3u
+
+// Peripheral interrupt ID ranges
+#define PLIC_FIRST_IRQ_ID kDifPlicGpio0
+#define PLIC_LAST_IRQ_ID kDifPlicNmiGenEsc3
+#define PLIC_GPIO_FIRST_IRQ_ID kDifPlicGpio0
+#define PLIC_GPIO_LAST_IRQ_ID kDifPlicGpio31
+#define PLIC_UART_FIRST_IRQ_ID kDifPlicUartTxWatermark
+#define PLIC_UART_LAST_IRQ_ID kDifPlicUartRxParityErr
+#define PLIC_SPI_DEVICE_FIRST_IRQ_ID kDifPlicSpiDeviceRxF
+#define PLIC_SPI_DEVICE_LAST_IRQ_ID kDifPlicSpiDeviceTxUnderflow
+#define PLIC_FLASH_CONTROL_FIRST_IRQ_ID kDifPlicFlashCtrlProgEmpty
+#define PLIC_FLASH_CONTROL_LAST_IRQ_ID kDifPlicFlashCtrlOpError
+#define PLIC_HMAC_FIRST_IRQ_ID kDifPlicHmacDone
+#define PLIC_HMAC_LAST_IRQ_ID kDifPlicHmacErr
+#define PLIC_ALERT_HANDLER_FIRST_IRQ kDifPlicAlertHandlerClassA
+#define PLIC_ALERT_HANDLER_LAST_IRQ kDifPlicAlertHandlerClassD
+#define PLIC_NMI_GEN_FIRST_IRQ kDifPlicNmiGenEsc0
+#define PLIC_NMI_GEN_LAST_IRQ kDifPlicNmiGenEsc3
+
+// Peripheral IRQ sources that interrupt the Ibex core
+#define PLIC_IBEX_SOURCE_FIRST_IRQ kDifPlicGpio0
+#define PLIC_IBEX_SOURCE_LAST_IRQ kDifPlicNmiGenEsc3
+
+// This define is used to calculate the IRQ index in IP, LE, IE registers.
+// These registers are 32bit wide, and in order to accommodate all the IRQs,
+// multiple of the same type registers are defined (IE00, IE01, ...). For
+// example, IRQ ID 33 corresponds to bit 1 in registers IP1, LE1, IE01. The
+// value 33 is chosen instead of 32 because IRQ IDs start from 1 rather than 0.
+#define PLIC_ID_TO_INDEX_MODULO 33u
+
+static bool plic_priority_validate(uint32_t priority) {
+  return priority <= RV_PLIC_MAX_PRIORITY;
+}
+
+/**
+ * Check the PLIC state for whether an IRQ has been claimed
+ */
+static bool plic_irq_claimed(dif_plic_t *plic) {
+  return (plic->claimed_irq.peripheral != kDifPlicPeripheralNone &&
+          plic->claimed_irq.source != kDifPlicNoIrq);
+}
+
+static void plic_irq_claimed_data_clear(dif_plic_t *plic) {
+  plic->claimed_irq.peripheral = kDifPlicPeripheralNone;
+  plic->claimed_irq.source = kDifPlicNoIrq;
+}
+
+/**
+ * Get IE, IP or LE register offset (IE00, IE01, ...) from IRQ source ID.
+ */
+static ptrdiff_t plic_offset_from_reg0(dif_plic_irq_src_id_t irq) {
+  uint8_t register_index = (uint32_t)irq / PLIC_ID_TO_INDEX_MODULO;
```

Comment:
This breaks if we ever support more than 64 sources. Should be `(irq - 1) / 32` to be correct in general.